---
title: "Illustration of credible set using Matthew's \"proxy SNP\" idea"
author: "Peter Carbonetto"
date: October 19, 2017
output:
  html_document:
    theme: readable
    include:
      before_body: include/header.html
      after_body: include/footer.html
---

```{r knitr, echo=FALSE}
knitr::opts_chunk$set(
  comment   = "#",
  results   = "hold",
  collapse  = TRUE,
  fig.align = "center")
```

## Outline

The aim of this analysis is to try out a simple fine-mapping idea.

## Set up environment

Load a few packages.

```{r}
library(ggplot2)
library(cowplot)
library(varbvs)
```

Also, install the varbvs package from the "finemap" branch on Github.

```{r, eval=FALSE}
library(devtools)
install_github("pcarbo/varbvs",ref = "finemap",subdir = "varbvs-R")
```

## Script parameters

This vector specifies the (additive) phenotypic effects of the
SNPs. In this case, exactly two SNPs (about 3 Mb apart on chromosome
11) affect the quantitative trait.

```{r}
beta            <- rep(0,531)
beta[c(88,506)] <- c(0.2,-0.35)
```

```{r}
ld.threshold <- 0.2
```

## Load genotype data

Load the genotype data for the 531 SNPs at the bone-mineral density
locus on chromosome 11 that was identified in the *Nature Genetics*
manuscript.

These are data from outbred (CFW) mice, so will not have the same LD
patterns as humans. I'm using mouse instead of human data because we
don't have to worry about about making the mouse data publicly
available.

```{r}
load("../datafiles/cfw.chr11bmd.RData")
n <- nrow(X)
p <- ncol(X)
```

Center the columns of X so that each column has a mean of zero.

```{r}
rep.row <- function (x, n)
  matrix(x,n,length(x),byrow = TRUE)
X <- X - rep.row(colMeans(X),n)
```

## Simulate phenotype data

Set the random number generator seed.

```{r}
set.seed(1)
```

Generate the quantitative trait measurements.

```{r}
y <- c(X %*% beta + rnorm(n))
```

## Fit variational approximation to posterior

Fit the fully-factorized variational approximation to the posterior
distribution of the coefficients for a linear regression model of a
continuous outcome (quantitiative trait), with spike and slab priors on
the coefficients. 

```{r}
fit <- varbvs(X,NULL,y,sa = 1,verbose = FALSE)
```

## Compute proxy probabilities

Compute the correlations between all the SNPs.

```{r}
R <- cor(X)
```

To test the idea of computing credible SNP sets based on "proxy
probabilities", here I choose the top two SNPs by posterior inclusion
probability (PIP), take all candidate SNPs in LD with these top two
SNPs (based on a pre-specified threshold), then compute the Bayes
factors measuring the improvement in "fit" by using the candidate SNPs
instead of each of the top SNPs.

```{r}
top.markers <- summary(fit)$top.vars$index
i     <- top.markers[1]
j     <- top.markers[2]
vars1 <- which(R[,i]^2 >= ld.threshold)
vars2 <- which(R[,j]^2 >= ld.threshold)
BF1   <- varbvsproxybf(X,NULL,y,fit,i,vars1) 
BF2   <- varbvsproxybf(X,NULL,y,fit,j,vars2)
```

The 

```{r}
bf1 <- BF1[,1]
bf2 <- BF2[,1]
pp1 <- bf1/sum(bf1)
pp2 <- bf2/sum(bf2)
```

Explain what this code chunk does.

```{r}
markers1 <- order(bf1,decreasing = TRUE)
n1       <- min(which(cumsum(pp1[markers1]) > 0.95))
markers1 <- markers1[1:n1]
```

TO DO: Explain what this code chunk does.

```{r}
markers2 <- order(bf2,decreasing = TRUE)
n2       <- min(which(cumsum(pp2[markers2]) > 0.95))
markers2 <- markers2[1:n2]
```

## Plot fine-mapping results

TO DO: Explain what this plot shows.

```{r}
clrs <- c("midnightblue","darkviolet","darkorchid","maroon",
          "tomato","orange","gold","yellow")
dat1 <- data.frame(pos = 1:p,pip = fit$pip)
regionplot1 <- ggplot(mapping = aes(x = pos,y = pip,color = abs(R[,i]))) +
  geom_point(data = dat1,size = 2,shape = 19) +
  geom_point(data = dat1[vars1[markers1],],color = "black",shape = 4,
             size = 0.5) +
  scale_x_continuous(breaks = NULL) +
  scale_y_continuous(trans = "log10",breaks = 10^(-4:0)) +
  scale_color_gradientn(colors = clrs) +
  labs(x = "position on chromosome 11, 94-98 Mb",y = "varbvs PIP",
       title = paste("correlations with",map[i,"id"])) +
  theme_cowplot(font_size = 11) +
  theme(axis.line = element_blank())
```

TO DO: Explain what this plot shows.

```{r}
dat2 <- data.frame(pos = 1:p,pip = fit$pip)
regionplot2 <- ggplot(mapping = aes(x = pos,y = pip,color = abs(R[,j]))) +
  geom_point(data = dat2,size = 2,shape = 19) +
  geom_point(data = dat2[vars2[markers2],],color = "black",shape = 4,
             size = 0.5) +
  scale_x_continuous(breaks = NULL) +
  scale_y_continuous(trans = "log10",breaks = 10^(-4:0)) +
  scale_color_gradientn(colors = clrs) +
  labs(x = "position on chromosome 11, 94-98 Mb",y = "varbvs PIP",
       title = paste("correlations with",map[j,"id"])) +
  theme_cowplot(font_size = 11) +
  theme(axis.line = element_blank())
```

Arrange the plots.

```{r}
print(plot_grid(regionplot1,regionplot2,nrow = 2))
```
