---
title: "Illustration of Matthew's \"proxy SNP\" idea on CFW data set"
author: "Peter Carbonetto"
date: October 20, 2017
output:
  html_document:
    theme: readable
    include:
      before_body: include/header.html
      after_body: include/footer.html
---

```{r knitr, echo=FALSE}
knitr::opts_chunk$set(
  comment   = "#",
  results   = "hold",
  collapse  = TRUE,
  fig.align = "center")
```

## Outline

The aim of this analysis is to try out a simple fine-mapping idea.  We
can also build on this analysis to test out other ideas for
fine-mapping (that is, estimating credible sets for the location of
the causal variant near an identified assocation).

## Set up environment

Load a few packages.

```{r}
library(ggplot2)
library(cowplot)
library(varbvs)
```

Also, install the varbvs package from the "finemap" branch on Github.

```{r, eval=FALSE}
library(devtools)
install_github("pcarbo/varbvs",ref = "finemap",subdir = "varbvs-R")
```

## Script parameters

This vector specifies the (additive) phenotypic effects of the
SNPs. In this case, exactly 2 SNPs (about 3 Mb apart on chromosome
11) affect the quantitative trait.

```{r}
beta            <- rep(0,531)
beta[c(88,506)] <- c(0.2,-0.35)
```

The causal variant should be within the credible set with this
probability.

```{r}
credint.prob <- 0.95
```

For identifying the credible set, compute proxy probabilities for all
SNPs in LD with the target SNP based on this threshold.

```{r}
ld.threshold <- 0.2
```

## Load genotype data

Load the genotype data for the 531 SNPs at the bone-mineral density
locus on chromosome 11 that was identified in the *Nature Genetics*
manuscript.

These are data from outbred (CFW) mice, so will not have the same LD
patterns as humans. I'm using mouse instead of human data because we
don't have to worry about about making the mouse data publicly
available.

```{r}
load("../datafiles/cfw.chr11bmd.RData")
n <- nrow(X)
p <- ncol(X)
```

Center the columns of X so that each column has a mean of zero.

```{r}
rep.row <- function (x, n)
  matrix(x,n,length(x),byrow = TRUE)
X <- X - rep.row(colMeans(X),n)
```

## Simulate phenotype data

Set the random number generator seed.

```{r}
set.seed(1)
```

Generate the quantitative trait measurements.

```{r}
y <- c(X %*% beta + rnorm(n))
```

## Fit variational approximation to posterior

Fit the fully-factorized variational approximation to the posterior
distribution of the coefficients for a linear regression model of a
continuous outcome (quantitiative trait), with spike and slab priors on
the coefficients. 

```{r}
fit <- varbvs(X,NULL,y,sa = 1,verbose = FALSE)
```

## Compute proxy probabilities

Compute the correlations between all the SNPs.

```{r}
R <- cor(X)
```

To test the idea of computing credible SNP sets based on "proxy
probabilities", here I choose the top 2 SNPs by posterior inclusion
probability (PIP), take all candidate SNPs in LD with these top 2 SNPs
(based on a pre-specified threshold), then compute the Bayes factors
measuring the improvement in "fit" by using the candidate SNPs instead
of each of the top SNPs.

First, do this for the top SNP.

```{r}
top.markers <- summary(fit)$top.vars$index
i     <- top.markers[1]
vars1 <- which(R[,i]^2 >= ld.threshold)
BF1   <- varbvsproxybf(X,NULL,y,fit,i,vars1) 
```

Note that the Bayes factors are computed separately for each
hyperparameter setting. Here, to simplify the example, we will just
use the Bayes factors from the first hyperparameter setting
(corresponding to the smallest prior inclusion probability).

Also note that the implementation of `varbvsproxybf` is currently very
inefficient and probably will be too slow for examples with thousands
of SNPs.

Next, do the same for the SNP with the second-highest PIP.

```{r}
top.markers <- summary(fit)$top.vars$index
j     <- top.markers[2]
vars2 <- which(R[,j]^2 >= ld.threshold)
BF2   <- varbvsproxybf(X,NULL,y,fit,j,vars2)
```

This function computes the "probability probabilities" assuming that
exactly 1 SNP is causal, then selects the smallest set of SNPs such
that the sum of the proxy probabilities is greater than "pr".

```{r}
get.credible.set <- function (bf, pr) {
  pp      <- bf/sum(bf)
  markers <- order(bf,decreasing = TRUE)
  n       <- min(which(cumsum(pp[markers]) > credint.prob))
  return(markers[1:n])
}
```

Get the credible SNP sets for the top 2 SNPs (using the first
hyperparameter setting only).

```{r}
cred1 <- vars1[get.credible.set(BF1[,1],credint.prob)]
cred2 <- vars2[get.credible.set(BF2[,1],credint.prob)]
print(c(length(cred1),length(cred2)))
```

## Plot fine-mapping results

Plot the PIPs of the SNPs, marking the two credible SNP sets with
crosses, and colouring the points according to LD with the top 2 SNPs.

```{r, fig.height=6, fig.width=8}
create.finemap.plot <- function (pip, r, markers, plot.title = NULL) {
  clrs <- c("midnightblue","darkviolet","darkorchid","maroon",
            "tomato","orange","gold","yellow")
  dat <- data.frame(pos = 1:p,pip = fit$pip,r = r)
  return(ggplot(mapping = aes(x = pos,y = pip,color = r),
                environment = environment()) +
    geom_point(data = dat,size = 2,shape = 19) +
    geom_point(data = dat[markers,],color = "black",shape = 4,size = 0.5) +
    scale_x_continuous(breaks = NULL) +
    scale_y_continuous(trans = "log10",breaks = 10^(-4:0)) +
    scale_color_gradientn(colors = clrs,name = "abs(r)") +
    labs(x = "position on chromosome 11, 94-98 Mb",y = "varbvs PIP",
         title = plot.title) +
    theme_cowplot(font_size = 11) +
    theme(axis.line  = element_blank(),
	      plot.title = element_text(size = 11,face = "plain")))
}
regionplot1 <-
  create.finemap.plot(fit$pip,abs(R[,i]),cred1,
                      paste("colors indicate correlations with",map[i,"id"]))
regionplot2 <-
  create.finemap.plot(fit$pip,abs(R[,j]),cred2,
                      paste("colors indicate correlations with",map[j,"id"]))
plot_grid(regionplot1,regionplot2,nrow = 2)
```
