---
title: "Illustration of Matthew's \"proxy SNP\" idea on GTEx data set"
author: "Peter Carbonetto"
date: October 19, 2017
output:
  html_document:
    theme: readable
    include:
      before_body: include/header.html
      after_body: include/footer.html
---

```{r knitr, echo=FALSE}
knitr::opts_chunk$set(
  comment   = "#",
  results   = "hold",
  collapse  = TRUE,
  fig.align = "center")
```

## Outline

The aim of this analysis is to try out a simple fine-mapping idea.  We
can also build on this analysis to test out other ideas for
fine-mapping (that is, estimating credible sets for the location of
the causal variant near an identified assocation).

## Set up environment

Load a few packages.

```{r}
library(ggplot2)
library(cowplot)
library(varbvs)
```

Also, install the varbvs package from the "finemap" branch on Github.

```{r, eval=FALSE}
library(devtools)
install_github("pcarbo/varbvs",ref = "finemap",subdir = "varbvs-R")
```

## Script parameters

The causal variant should be within the credible set with this
probability.

```{r}
credint.prob <- 0.95
```

For identifying the credible set, compute proxy probabilities for all
SNPs in LD with the target SNP based on this threshold.

```{r}
ld.threshold <- 0.2
```

## Load genotype data

Load the GTEx genotype data for the 4,546 SNPs at ACTN3 locus.

```{r}
load("../datafiles/Muscle_Skeletal.ACTN3.1Mb.RData")
storage.mode(X) <- "double"
n <- nrow(X)
p <- ncol(X)
```

## Fit variational approximation to posterior

Fit the fully-factorized variational approximation to the posterior
distribution of the coefficients for a linear regression model of a
continuous outcome (quantitiative trait), with spike and slab priors on
the coefficients. 

```{r}
fit <- varbvs(X,Z,y,verbose = FALSE)
```

## Compute proxy probabilities

Compute the correlations between all the SNPs.

```{r}
R <- cor(X)
```

To test the idea of computing credible SNP sets based on "proxy
probabilities", here I choose the top SNP by posterior inclusion
probability (PIP), take all candidate SNPs in LD with this top SNP
(based on a pre-specified threshold), then compute the Bayes factors
measuring the improvement in "fit" by using the candidate SNPs instead
of the top SNP.

```{r}
top.markers <- summary(fit)$top.vars$index
i    <- top.markers[1]
vars <- which(R[,i]^2 >= ld.threshold)
out  <- varbvsproxybf(X,Z,y,fit,i,vars)
```

Note that the Bayes factors are computed separately for each
hyperparameter setting. Here, to simplify the example, we will just
use the Bayes factors from the first hyperparameter setting
(corresponding to the smallest prior inclusion probability).

This function computes the "probability probabilities" assuming that
exactly 1 SNP is causal, then selects the smallest set of SNPs such
that the sum of the proxy probabilities is greater than "pr".

```{r}
get.credible.set <- function (bf, pr) {
  pp      <- bf/sum(bf)
  markers <- order(bf,decreasing = TRUE)
  n       <- min(which(cumsum(pp[markers]) > credint.prob))
  return(markers[1:n])
}
```

Get the credible SNP set for the top SNP (using the first
hyperparameter setting only).

```{r}
cr <- vars[get.credible.set(out$BF[,1],credint.prob)]
```

## Plot fine-mapping results

Plot the PIPs of the SNPs, marking the two credible SNP sets with
crosses, and colouring the points according to LD with the top 2 SNPs.

```{r, fig.height=6, fig.width=8}
create.finemap.plot <- function (pip, r, markers, plot.title = NULL) {
  clrs <- c("midnightblue","darkviolet","darkorchid","maroon",
            "tomato","orange","gold","yellow")
  dat <- data.frame(pos = 1:p,pip = fit$pip,r = r)
  return(ggplot(mapping = aes(x = pos,y = pip,color = r),
                environment = environment()) +
    geom_point(data = dat,size = 2,shape = 1) +
    geom_point(data = dat[markers,],color = "black",shape = 4,size = 0.5) +
    scale_x_continuous(breaks = NULL) +
    scale_y_continuous(trans = "log10",breaks = 10^(-4:0)) +
    scale_color_gradientn(colors = clrs,name = "abs(r)") +
    labs(x = "position on chromosome 11, 94-98 Mb",y = "varbvs PIP",
         title = plot.title) +
    theme_cowplot(font_size = 11) +
    theme(axis.line  = element_blank(),
	      plot.title = element_text(size = 11,face = "plain")))
}
print(create.finemap.plot(fit$pip,abs(R[,i]),cr,
      paste("colors indicate correlations with",colnames(X)[i])))
```
