---
title: "Illustration of SuSie model for fine-mapping"
author: "Peter Carbonetto"
date: November 28, 2017
output:
  html_document:
    theme: readable
    include:
      before_body: header.html
---

```{r knitr, echo=FALSE}
knitr::opts_chunk$set(
  comment   = "#",
  results   = "hold",
  collapse  = TRUE,
  fig.align = "center")
```

This is a small example to compare the fully-factorized variational
approximation against the SuSiE ("Sum of Single Effects") approach for
fine-mapping. Here, the phenotype is expression of the *FMO2*
gene. The data are genotypes from the GTEx study, and simulated
phenotypes.

## Script parameters

```{r set-params}
k     <- 3       # Number of nonzero effects in model
sigma <- 9       # Variance of residual.
sa    <- 2       # Prior variance of nonzero effects.
pp    <- 3/1000  # Prior inclusion probability.
```

These are the additive effects used to simulate the
phenotype data.

```{r set-beta}
beta                 <- rep(0,1000)
beta[c(201,562,952)] <- c(3,-3,5)
```

## Set up environment

Load some functions and packages used in this example.

```{r setup}
library(varbvs)
library(ggplot2)
library(cowplot)
source("varbvssusie.R")
```

Set the sequence of pseudorandom numbers (useful for making sure
that the results are reproducible).

```{r set-seed}
set.seed(1)
```

## Load genotype data

Retrieve genotype data for 1,000 SNPs near gene *FMO2*.

```{r load-data}
load("../datafiles/Thyroid.FMO2.1Mb.RData")
rm(Z,y)
X <- X[,3501:4500]
n <- nrow(X)
p <- ncol(X)
storage.mode(X) <- "double"
```

Center the columns of the genotype matrix.

```{r center-genotypes}
X <- scale(X,center = TRUE,scale = FALSE)
```

## Simulate phenotype data

I also center the phenotypes.

```{r sim-data}
y <- c(X %*% beta + sqrt(sigma)*rnorm(n))
y <- y - mean(y)
```

## Fit fully-factorized variational approximation

Note that changing the update order will result in very different results.

```{r fit-varbvs} 
alpha0  <- runif(p)
alpha0  <- alpha0/sum(alpha0)
mu0     <- rnorm(p)
pp      <- rep(pp,1,p)
logodds <- varbvs:::logit(pp)
fit1    <- varbvsnorm(X,y,sigma,sa,logodds,alpha0,mu0,update.order = 1:p,
                      update.sigma = FALSE,update.sa = FALSE,tol = 1e-6,
					  verbose = FALSE)
```

## Fit "SuSiE" variational approximation

We see that the co-ordinate ascent updates converge to a stationary
point more rapidly than the updates for the fully-factorized
approximation.

```{r fit-susie}
fit2 <- varbvssusie(X,y,k,sigma,sa,pp,tol = 1e-6,maxiter = 100)
```

## Summarize results

Not sure it is appropriate to do so, but I thought it might be
interesting to compare the "fit" of the two models by comparing
the variational lower bounds to the marginal likelihoods.

```{r compare-likelihoods}
cat("Approximate marginal log-likelihoods:\n")
last <- function (x) x[length(x)]
cat(sprintf("varbvsnorm  = %0.6f\n",last(fit1$logw)))
cat(sprintf("varbvssusie = %0.6f\n",last(fit2$logw)))
cat(sprintf("Improvement in likelihood using SuSie model: %0.2e\n",
            exp(last(fit2$logw) - last(fit1$logw))))
```

Show convergence of the co-ordinate ascent updates for the two
methods.

```{r plot-progress, fig.height=3.5, fig.width=7}
plot.progress <- function (logw, clr, plot.title)
  ggplot(data.frame(x = 1:length(logw),
                    y = log10(max(logw) - logw)),
		 aes(x = x,y = y)) +
  geom_line(color = clr,size = 1) +
  labs(x = "iteration",y = "log10-distance from max",title = plot.title)
p1 <- plot.progress(fit1$logw,"dodgerblue","fully-factorized")
p2 <- plot.progress(fit2$logw,"darkorange","sum of single effects")
plot_grid(p1,p2)
```

We observe that both algorithms progress rapidly to a fixed point, at
least in this example.

Next, I plot the posterior inclusion probabilities (PIPs) under the
fully-factorized variational approximation, and under the SuSiE
model. For the SuSiE model, I show, for each SNP, the posterior
probability that one or more of the coefficients is nonzero, which is
easily calculated under the variational approximation.

```{r plot-pips, fig.width=8, fig.height=6}
plot.pips <- function (pip, r2, plot.title) {
  n    <- length(pip)
  clrs <- c("#0D0887FF","#5402A3FF","#8B0AA5FF","#B93289FF",
            "#DB5C68FF","#F48849FF","#FEBC2AFF","#F0F921FF")
  return(ggplot(data.frame(x = 1:p,y = pip,r2 = r2),
             aes(x = x,y = y,color = r2)) +
    geom_point(shape = 20,size = 2) +
    scale_x_continuous(breaks = NULL) +
    scale_color_gradientn(colors = clrs) +
    labs(x = "",y = "PIP",title = plot.title))
}
markers <- which(beta != 0)
r2  <- apply((cor(X)^2)[,markers],1,max)
pip <- 1 - apply(1 - fit2$alpha,1,prod)
p3  <- plot.pips(fit1$alpha,r2,"fully-factorized")
p4  <- plot.pips(pip,r2,"sum of single effects") +
         geom_point(data = data.frame(x = markers,y = pip[markers]),
                    aes(x = x,y = y),color = "black",shape = 4,
					size = 0.75)
print(plot_grid(p3,p4,nrow = 2))
```

Additionally, to understand how the the SuSiE model captures
uncertainty in the location of the SNP effect, I used colour to show
correlation strength (LD) with the true effects. In the lower panel,
the SNPs with (true) effects on the phenotype are marked with x's.

## Session information

Here are details about the computing environment that I used to
generate these results, including the version of R that was used, and
the versions of the packages that were loaded into the R session.

```{r session-info}
sessionInfo()
```
