---
title: "Another illustration of Matthew's \"proxy SNP\" idea on GTEx data set"
author: "Peter Carbonetto"
date: October 26, 2017
output:
  html_document:
    theme: readable
    include:
      before_body: include/header.html
      after_body: include/footer.html
---

```{r knitr, echo=FALSE}
knitr::opts_chunk$set(
  comment   = "#",
  results   = "hold",
  collapse  = TRUE,
  fig.align = "center")
```

## Outline

In this example, we consider fine-mapping for a more complicated
scenario in which multiple SNPs are independently associated with a
(gene expression) quantitative trait within a single locus.

## Set up environment

Load a few packages.

```{r, message=FALSE}
library(ggplot2)
library(cowplot)
library(varbvs)
```

Also, install the varbvs package from the "finemap" branch on Github.

```{r, eval=FALSE}
library(devtools)
install_github("pcarbo/varbvs",ref = "finemap",subdir = "varbvs-R")
```

## Script parameters

The causal variant should be within the credible set with this
probability.

```{r}
credint.prob <- 0.95
```

For identifying the credible set, compute proxy probabilities for all
SNPs in LD with the target SNP based on this threshold.

```{r}
ld.threshold <- 0.2
```

## Load genotype data

Retrieve genotype data for 7,651 SNPs in the *FMO2* locus.

```{r}
load("../datafiles/Thyroid.ENSG00000237846.1Mb.RData")
storage.mode(X) <- "double"
n <- nrow(X)
p <- ncol(X)
```

## Fit variational approximation to posterior

Fit the fully-factorized variational approximation to the posterior
distribution of the coefficients for a linear regression model of a
continuous outcome (quantitiative trait), with spike and slab priors
on the coefficients. We find that several SNPs in this locus are
independently associated with gene expression, even when the prior is
very conservative (1 SNP is associated with the gene expression trait
*a priori*).

```{r}
fit <- varbvs(X,Z,y,verbose = FALSE)
sum(fit$alpha[,1] > 0.95)
```

## Compute proxy probabilities

This function computes the "probability probabilities" assuming that
exactly 1 SNP is causal, then selects the smallest set of SNPs such
that the sum of the proxy probabilities is greater than "pr".

```{r}
get.credible.set <- function (bf, pr) {
  pp      <- bf/sum(bf)
  markers <- order(bf,decreasing = TRUE)
  n       <- min(which(cumsum(pp[markers]) > credint.prob))
  return(markers[1:n])
}
```

## Plot fine-mapping results

To plot the fine-mapping results, I highlight the credible SNP sets in
green , and otherwise points are coloured according to LD with each of
the top SNPs.

```{r}
create.finemap.plot <- function (pip, r, markers, plot.title = NULL) {
  clrs <- c("midnightblue","darkviolet","darkorchid","maroon",
            "tomato","orange","gold","yellow")
  dat <- data.frame(pos = 1:p,pip = fit$pip,r = r)
  return(ggplot(mapping = aes(x = pos,y = pip,color = r),
                environment = environment()) +
    geom_point(data = dat,size = 1.5,shape = 20) +
    geom_point(data = dat[markers,],color = "limegreen",
	           shape = 20,size = 1.5) +
    scale_x_continuous(breaks = NULL) +
    scale_y_continuous(trans = "log10",breaks = 10^(-4:0)) +
    scale_color_gradientn(colors = clrs,guide = FALSE) +
    labs(x = "",y = "varbvs PIP",title = plot.title) +
    theme_cowplot(font_size = 7) +
    theme(axis.line  = element_blank(),
	      plot.title = element_text(size = 7,face = "plain")))
}
```

Compute the correlations between all the SNPs.

```{r}
R <- cor(X)
```

To test the idea of computing credible SNP sets based on "proxy
probabilities", here I choose the top SNPs by posterior inclusion
probability (PIP), take all candidate SNPs in LD with each of the top
SNPs (based on a pre-specified threshold), then compute the Bayes
factors measuring the improvement in "fit" by using alternative SNPs
instead of the top SNP.

Get the SNPs with PIP > 0.95.

```{r}
top.markers <- summary(fit,nv = 20)$top.vars
top.markers <- subset(top.markers,prob > 0.95)$variable
m           <- length(top.markers)
```

Compute the credible SNP sets, taking the strategy as described
above. Note that the credible set may not include the top SNP---this
could happen because the Bayes factor calculations do not control for
the effects of nearby, independently associated SNPs as effectively as
the multi-SNP regression analysis (as implemented in `varbvs`).

```{r}
cr           <- vector("list",m)
plots        <- vector("list",m)
names(cr)    <- top.markers
names(plots) <- top.markers
for (i in top.markers) {
  idx        <- which(colnames(X) == i)
  vars       <- which(R[,i]^2 >= ld.threshold)
  out        <- varbvsproxybf(X,Z,y,fit,idx,vars)
  cr[[i]]    <- vars[get.credible.set(out$BF[,1],credint.prob)]
  j          <- names(which.max(out$BF[,1]))
  plots[[i]] <- create.finemap.plot(fit$pip,abs(R[,j]),cr[[i]],
                                    sprintf("%s (%d)",j,length(cr[[i]])))
}
```

Also note that the Bayes factors are computed separately for each
hyperparameter setting. Here, to simplify the example, I've taken the
Bayes factors from the first hyperparameter setting (corresponding to
the smallest prior inclusion probability).

```{r, fig.height=11, fig.width=8}
do.call("plot_grid",c(plots,list(nrow = 6,ncol = 2)))
```

In these plots, the number in the parentheses gives the number of SNPs
in the credible SNP set.


## Session information

This is the computing environment, version of R and packages that were
used to generate these results.

```{r session-info}
sessionInfo()
```
